curvlinops/kfac.py

```diff
@@ -261,12 +266,57 @@ def __init__(
             params,
             data,
             progressbar=progressbar,
-            check_deterministic=check_deterministic,
+            check_deterministic=False,
             shape=shape,
             num_data=num_data,
             batch_size_fn=batch_size_fn,
         )
 
+        self._set_num_per_example_loss_terms(num_per_example_loss_terms)
+
+        if check_deterministic:
+            old_device = self._device
+            self.to_device(device("cpu"))
+            try:
+                self._check_deterministic()
+            except RuntimeError as e:
+                raise e
+            finally:
+                self.to_device(old_device)
+
+    def _set_num_per_example_loss_terms(
+        self, num_per_example_loss_terms: Optional[int]
+    ):
+        """Set the number of per-example loss terms.
+
+        Args:
+            num_per_example_loss_terms: Number of per-example loss terms. If ``None``,
+                it is inferred from the data at the cost of one traversal through the
+                data loader.
+
+        Raises:
+            ValueError: If the number of loss terms is not divisible by the number of
+                data points.
+        """
+        if num_per_example_loss_terms is None:
+            # Determine the number of per-example loss terms
+            num_loss_terms = sum(
+                (
+                    y.numel()
+                    if isinstance(self._loss_func, CrossEntropyLoss)
+                    else y.shape[:-1].numel()
+                )
+                for (_, y) in self._loop_over_data(desc="_num_per_example_loss_terms")
+            )
+            if num_loss_terms % self._N_data != 0:
```

Comment:
Q: Should we `assert` here that all `y`s are of identical shape except for the batch dimension?

Reply:
The reason I don't do this is that then I have to assume that the batch dimension is the first one  (or some other one) -- maybe not a bad assumption, but we have a test (`test_multi_dim_output`) that relies on being able to use flattened labels. This is also the main reason I don't just determine `num_per_example_loss_terms` from the targets of a single batch.

Not really sure right now what is best here, so let me know if you have a preferred solution.

Reply:
I don't have a better solution at the moment.